client/packages/system/src/Stock/api/hooks/useStockLine.ts

```diff
@@ -0,0 +1,92 @@
+import { useCallback, useState } from 'react';
+import {
+  FnUtils,
+  setNullableInput,
+  useMutation,
+} from '@openmsupply-client/common';
+import {
+  InventoryAdjustmentReasonRowFragment,
+  StockLineRowFragment,
+} from '../../..';
+import { STOCK_LINE } from './keys';
+import { useStockGraphQL } from '../useStockGraphQL';
+
+export interface DraftStockLine extends StockLineRowFragment {
+  inventoryAdjustmentReason: InventoryAdjustmentReasonRowFragment | null;
+}
+
+export function useStockLine() {
+  const [draft, setDraft] = useState<DraftStockLine>({
+    __typename: 'StockLineNode',
+    id: '',
+    itemId: '',
+    onHold: false,
+    packSize: 0,
+    sellPricePerPack: 0,
+    costPricePerPack: 0,
+    totalNumberOfPacks: 0,
+    availableNumberOfPacks: 0,
+    storeId: '',
+    item: {
+      __typename: 'ItemNode',
+      code: '',
+      name: '',
+    },
+    inventoryAdjustmentReason: null,
```

Comment:
Keen for the thoughts here...

From the perspective of our new API hook design, I liked the idea of managing input state from the same hook that calls the mutation. However, if the UI needs a slightly different shape than the API, where would we do the mapping?

In this case, the draft state shape is a StockLineRowFragment (which has some extra stuff than needed for this API call, but that's the type needed for the stock edit form) plus the whole inventoryAdjustmentReason (we have to manage the whole thing in state for the `InventoryAdjustmentReasonSearchInput`… would be nicer if we only had to manage the id, but changes would also affect stocktake so I am refraining at this point lol)

ANYWAY this state should probably be closer to the components right? I think at that point, managing/exposing state from this hook wouldn't have a whole lot of value?

Reply:
I think it's fine how you've done it here. It would only be an issue if there was multiple usages of the same hook that required different state structures. But that isn't really the case in most situations, is it? In the event we find one, we can figure out the best way to handle it based on the specific case, perhaps? Can you think of any?

Reply:
A agree I like it, although I think we could remove things from the common type (like storeId, itemId etc..)